python/hail/history.py

```diff
@@ -0,0 +1,170 @@
+from __future__ import print_function  # Python 2 and 3 print compatibility
+
+import datetime
+import inspect
+import itertools
+from collections import OrderedDict
+from decorator import decorator
+from hail.utils import hadoop_write, wrap_to_list
+from hail.java import Env
+
+
+def parse_args(f, args, kwargs):
+    argspec = inspect.getargspec(f)
+    kwargs_not_default = OrderedDict({})
+
+    if argspec.defaults:
+        n_postnl_args = len(argspec.args) - len(argspec.defaults)
+        kwargs_not_default.update(OrderedDict({k:v for k, (v, d) in zip(argspec.args[n_postnl_args:], zip(args[n_postnl_args - 1:], argspec.defaults)) if v != d}))
+        postnl_args = args[:n_postnl_args - 1]
+    else:
+        postnl_args = args
+
+    kwargs_not_default.update(OrderedDict({k:v for k,v in kwargs.iteritems()}))
+
+    return postnl_args, kwargs_not_default
+
+
+@decorator
+def record_init(func, obj, *args, **kwargs):
+    postnl_args, kwargs_not_default = parse_args(func, args, kwargs)
+    h = History.from_init(type(obj).__name__, postnl_args, kwargs_not_default)
+    func(obj, *args, **kwargs)
+    obj._set_history(h)
+
+
+@decorator
+def record_method(func, obj, *args, **kwargs):
+    postnl_args, kwargs_not_default = parse_args(func, args, kwargs)
+
+    def set_history(item, index=None, key_name=None):
+        if isinstance(item, HistoryMixin):
+            item._set_history(obj
+                              ._get_history()
+                              .add_method(func.__name__, postnl_args, kwargs_not_default, index=index, key_name=key_name))
+
+    result = func(obj, *args, **kwargs)
+    if isinstance(result, dict):
+        for k, r in result.iteritems():
+            set_history(r, key_name=k)
+    elif isinstance(result, list) or isinstance(result, tuple):
+        for i, r in enumerate(result):
+            set_history(r, index=i)
+    else:
+        set_history(result)
+    return result
+
+
+@decorator
+def record_classmethod(func, cls, *args, **kwargs):
+    postnl_args, kwargs_not_default = parse_args(func, args, kwargs)
```

Comment:
wouldn't it be simpler to just write out method calls as kwargs all the time? I think that's equally clear or clearer to read as well.

e.g.
```python
hc.read_table(path='foo.tsv', impute=True)
```
instead of 
```python
hc.read_table('foo.tsv', impute=True)
```